File: frontend/packages/domain/format/src/model/masterKey.ts
import { drawBytes } from '../internal/entropy.js';
import { SecretBytes } from './secretBytes.js';

/** Length of a Master Key in bytes. */
export const MASTER_KEY_LENGTH = 32;

/**
 * The 256-bit key every purpose key in a Library is derived from.
 *
 * It is drawn from a CSPRNG and never from the Passphrase or any other
 * user-chosen input (KD-1), so the strength of the ciphertext on Storage never
 * depends on passphrase quality. Each Master Key epoch draws its own.
 */
export class MasterKey extends SecretBytes {
  private constructor(bytes: Uint8Array) {
    super(bytes, MASTER_KEY_LENGTH, 'MasterKey');
  }

  /** Takes 32 raw bytes. */
  static fromBytes(bytes: Uint8Array): MasterKey {
    return new MasterKey(bytes);
  }
}

/** Draws a fresh Master Key from the platform CSPRNG. */
export function generateMasterKey(): MasterKey {
  return MasterKey.fromBytes(drawBytes(MASTER_KEY_LENGTH));
}

File: backend/crates/domain/coffret-format/src/master_key.rs
use coffret_model::MasterKey;

use crate::entropy;
use crate::error::Result;

/// Draws a fresh Master Key from the operating system's CSPRNG.
///
/// The generator takes no user input at all — in particular not the Passphrase
/// — so the strength of the ciphertext on Storage never depends on passphrase
/// quality. Each Master Key epoch calls this again for a key of its own.
pub fn generate_master_key() -> Result<MasterKey> {
    Ok(MasterKey::from_bytes(entropy::draw()?))
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashSet;

    // KD-1: the Master Key is 256 bits drawn from the OS CSPRNG, and each epoch
    // draws its own.
    #[test]
    fn draws_distinct_256_bit_keys() {
        let keys: HashSet<[u8; MasterKey::BYTE_LEN]> = (0..256)
            .map(|_| {
                *generate_master_key()
                    .expect("the OS CSPRNG is available")
                    .as_bytes()
            })
            .collect();
        assert_eq!(keys.len(), 256);
        for key in &keys {
            assert_eq!(key.len(), 32);
        }
    }
}
